js/plugins/api.js

```diff
@@ -0,0 +1,213 @@
+            url  = `${this.root}/${url}`;
+        }
+        if (params && Object.keys(params).length) {
+            const parts = [url, this.encodeParams(params)];
+            url = url.indexOf('?') >= 0 ? parts.join('&') : parts.join('?');
+        }
+        return url;
+    }
+
+    /**
+     * Default handler for JSON response
+     * Parse the response if OK otherwise throw it
+     * (let the developper handle it in the catch() handler)
+     * @param  {Response} response A response with JSON expected body.
+     * @return {Promise}           A promise for data handling
+     */
+    onResponse(response) {
+        if (response.ok) {
+            return response.json();
+        } else {
+            return response.json()
+                .catch(() => Promise.reject(new ApiError(response)))
+                .then(json => Promise.reject(new ApiError(response, json)));
+        }
+    }
+
+
+    /**
+     * Generic error handler
+     * @param  {[type]} error [description]
+     * @return {[type]}       [description]
+     */
+    onError(error) {
+        if (error instanceof Response) {
+            error = new ApiError(response);
+        }
+        log.error('Unhandled API error:', error);
+        return Promise.reject(error);
+    }
+
+    /**
+     * Perform a GET operation
+     * @param  {String} url     An absolute or API root relative URL
+     * @param  {Object} params  Optionnal query strings parameters
+     * @param  {[type]} headers Optionnal HTTP headers
+     * @return {Promise}        An instanciated fetch promise
+     */
+    get(url, params, headers) {
+        return fetch(this.build(url, params), {
+                method: 'get',
+                credentials: 'include',
+                headers: Object.assign({}, DEFAULT_HEADERS, headers || {})
+            })
+            .catch(this.onError)
+            .then(this.onResponse);
+    }
+
+    /**
+     * Perform a POST operation with a JSON payload
+     * @param  {String} url     An absolute or API root relative URL
+     * @param  {Object} data    The body payload
+     * @param  {[type]} headers Optionnal HTTP headers
+     * @return {Promise}        An instanciated fetch promise
+     */
+    post(url, data, headers) {
+        return fetch(this.build(url), {
+            method: 'post',
+            credentials: 'include',
+            headers: Object.assign({}, DEFAULT_HEADERS, WRITE_HEADERS, headers || {}),
+            body: JSON.stringify(data)
+        })
+        .catch(this.onError)
+        .then(this.onResponse);
+    }
+
+    /**
+     * Perform a POST upload operation (multipart/form-data)
+     * @param  {String} url     An absolute or API root relative URL
+     * @param  {Object} data    The body payload
+     * @param  {Object} headers Optionnal HTTP headers
+     * @return {Promise}        An instanciated fetch promise
+     */
+    upload(url, data, headers) {
+        const body = data; // TODO handle body
```

Comment:
Interesting.
We also have FineUploader as dependency, used both in front and admin.
It was chosen for resumable chunked upload but this feature is not yet used (should allow big file upload).

I think upload experience is a key feature that need it's own PR (and right now it's only used in admin)
In this PR, I haven't the need for upload method so I left a TODO. Maybe I should throw an error instead or remove the method